### PR TITLE
Add PayPal button to bundles landing page

### PR DESCRIPTION
**Pull request by desbo**
## Why are you doing this?
So that people can make one-off contributions with PayPal without navigating to another page. 

pp

p.s. this uses `alert` to display PayPal errors, the most likely cause being the request to contributions failing (which is unlikely). Although it seems strange to use `alert` I've chatted to a few people and have decided it's not worth implementing a modal dialog on this page to show this single error.

### CODE REVIEW

File: app/views/react.scala.html [hunk 1]
```diff
@@ -1,8 +1,18 @@
 @import assets.AssetsResolver
-@(title: String, id: String, js: String, description: Option[String] = None)(implicit assets: AssetsResolver, request: RequestHeader)
+@(
+    title: String,
+    id: String,
+    js: String,
+    contributionsPayPalEndpoint: String,
```
Comment (svillafe) [inline app/views/react.scala.html]: The `react.scala.html` template is a generic template for pages that run react. Since it is generic, I would not add the contributionPayPalEndpoint to it. I would create a new template for the bundleLandingPage or a generic PayPalTemplate.

File: assets/components/payPalContributionButton/payPalContributionButton.jsx [hunk 2]
```diff
@@ -52,7 +53,7 @@ const PayPalContributionButton = (props: PropTypes) =>
   (
     <button
       id="qa-contribute-paypal-button"
-      className="component-paypal-contribution-button"
+      className={`component-paypal-contribution-button ${props.additionalClass || ''}`}
```
Comment (svillafe) [inline assets/components/payPalContributionButton/payPalContributionButton.jsx]: Is it possible to use the helper defined [here](https://github.com/guardian/support-frontend/blob/master/assets/helpers/utilities.js#L22) to handle this case?

File: app/views/react.scala.html [hunk 3]
```diff
@@ -1,8 +1,16 @@
...
+    title: String,
+    id: String,
+    js: String,
+    description: Option[String] = None
+)(implicit assets: AssetsResolver, request: RequestHeader)
 
 @scripts = {
     <script type="text/javascript" src="@assets(js)"></script>
+    <script type="text/javascript">
+        window.guardian = window.guardian || {};
```
Comment (svillafe) [inline app/views/react.scala.html]: Do you still need the guardian object in this template?

Comment (desbo) [reply app/views/react.scala.html]: ah, nope!

File: assets/pages/bundles-landing/components/Bundles.jsx [hunk 4]
```diff
@@ -29,6 +30,7 @@ import { getSubsLinks } from '../helpers/externalLinks';
 
 import type { SubsUrls } from '../helpers/externalLinks';
 import type { Participations } from '../../../helpers/abtest';
+import type { ReferrerAcquisitionData } from '../../../helpers/tracking/acquisitions';
```
Comment (svillafe) [inline assets/pages/bundles-landing/components/Bundles.jsx]: I would change this import to be: 
`import type { ReferrerAcquisitionData } from 'helpers/tracking/acquisitions';`

In that way, it is clearer that is a _shared_ helper. Additionally, would you be able to update the `abtest` import?

Comment (desbo) [reply assets/pages/bundles-landing/components/Bundles.jsx]: Done. In dotcom I used `declare type` (in the `flow-typed` directory) for these kind of definitions, so they're available throughout the project - a similar thing could be useful here. 

https://github.com/guardian/frontend/blob/master/static/src/javascripts/__flow__/types/ab-tests.js

File: assets/helpers/payPalContributionsCheckout/payPalContributionsCheckout.js [hunk 5]
```diff
@@ -92,5 +92,5 @@ export function paypalContributionsRedirect(
       throw response;
     })
     .then((res) => { window.location = res.approvalUrl; })
-    .catch(() => errorHandler('Sorry, an error occurred, please try again or use another payment method.'));
+    .catch(errorHandler.bind(null, 'Sorry, an error occurred, please try again or use another payment method.'));
```
Comment (svillafe) [inline assets/helpers/payPalContributionsCheckout/payPalContributionsCheckout.js]: Just for curiosity, what is the `.bind(null, ` for?

Comment (desbo) [reply assets/helpers/payPalContributionsCheckout/payPalContributionsCheckout.js]: It's just an alternative to an arrow function - `bind` creates a new function, the first parameter is what the value of `this` will be set to there. If this isn't conventional for the project I can change it back to the arrow

File: assets/pages/bundles-landing/components/Bundles.jsx [hunk 6]
```diff
@@ -256,6 +290,20 @@ function ContributionBundle(props: PropTypes) {
...
         {...props}
       />
+
+      {props.contribType === 'ONE_OFF' && !!contribAttrs.paypalCta &&
+        <PayPalContributionButton
+          buttonText={contribAttrs.paypalCta.text}
+          amount={Number(props.contribAmount.oneOff.value)}
+          referrerAcquisitionData={props.referrerAcquisitionData}
+          isoCountry={props.isoCountry}
+          /* eslint no-alert: "off" */
```
Comment (svillafe) [inline assets/pages/bundles-landing/components/Bundles.jsx]: I would restrict the scope of this comment by using:

`// eslint-disable-next-line no-alert: "off"`

File: assets/components/payPalContributionButton/payPalContributionButton.scss [hunk 7]
```diff
@@ -42,9 +42,17 @@
...
     }
   }
+}
+
+.component-paypal-contribution-button--bundle-landing {
+  width: 100%;
+
+  &.contrib-error {
+    top: -16px;
+  }
```
Comment (Ap0c) [inline assets/components/payPalContributionButton/payPalContributionButton.scss]: Am I understanding this correctly that you only want to apply these styles when the button appears on the bundles landing page?

We've set up our CSS so that these shared components should ideally know nothing about the context in which they are used. Then context-specific styles are applied through the page-level stylesheets.

For example, in this case, this component is shared across both the bundles landing and contributions landing pages. Any styles that are generic, part of the component regardless of where it's used, sit in `payPalContributionButton.scss`. Then, any styles that want to modify it on the specific pages where it is used are placed in the stylesheets for those pages. The page stylesheets always sit below the component stylesheets in the load order, so page-level styles always override component-level styles.

To summarise my ramblings, I think these style rules should sit within `bundlesLanding.scss`, e.g. [here](https://github.com/guardian/support-frontend/blob/master/assets/pages/bundles-landing/bundlesLanding.scss#L217) 😊:

```scss
.component-paypal-contribution-button {
  width: 100%;

  &.contrib-error {
    top: -16px;
  }
}
```

File: assets/pages/bundles-landing/components/Bundles.jsx [hunk 8]
```diff
@@ -10,10 +10,14 @@ import type { ListItem } from 'components/featureList/featureList';
 import CtaLink from 'components/ctaLink/ctaLink';
 import Bundle from 'components/bundle/bundle';
 import ContribAmounts from 'components/contribAmounts/contribAmounts';
+import PayPalContributionButton from 'components/payPalContributionButton/payPalContributionButton';
 import type { Contrib, Amounts, ContribError } from 'helpers/contributions';
 import type { IsoCountry } from 'helpers/internationalisation/country';
 import type { IsoCurrency, Currency } from 'helpers/internationalisation/currency';
 import type { Campaign } from 'helpers/tracking/acquisitions';
+import type { Participations } from 'helpers/abtest';
```
Comment (Ap0c) [inline assets/pages/bundles-landing/components/Bundles.jsx]: Nice! Much cleaner than: `'../../../helpers/abtest`... 😯

File: assets/pages/bundles-landing/components/Bundles.jsx [hunk 9]
```diff
@@ -256,6 +290,20 @@ function ContributionBundle(props: PropTypes) {
...
+        <PayPalContributionButton
+          buttonText={contribAttrs.paypalCta.text}
+          amount={Number(props.contribAmount.oneOff.value)}
+          referrerAcquisitionData={props.referrerAcquisitionData}
+          isoCountry={props.isoCountry}
+          // eslint-disable-next-line no-alert
+          errorHandler={e => alert(e)}
+          abParticipations={props.abTests}
+          additionalClass={`bundle-landing ${props.contribError ? 'contrib-error' : ''}`}
+        />
```
Comment (Ap0c) [inline assets/pages/bundles-landing/components/Bundles.jsx]: As discussed offline, in the case where a user enters an invalid value into the 'other amount' box, this button will still be clickable, and will send a bad request to the contributions backend.

To prevent this I think you could just pass `canClick={!props.contribError}` and the button will take care of the rest.

File: assets/pages/bundles-landing/components/Bundles.jsx (hunk 8 above)
Comment (desbo) [reply assets/pages/bundles-landing/components/Bundles.jsx]: I think intellij adds that!

File: assets/components/payPalContributionButton/payPalContributionButton.scss (hunk 7 above)
Comment (desbo) [reply assets/components/payPalContributionButton/payPalContributionButton.scss]: Thanks Jamie - I've moved that around and it's much clearer.

**Comment by desbo** [comment]
Button order swapped at Mario's request:

screen shot 2017-11-08 at 11 51 29
screen shot 2017-11-08 at 11 51 34
screen shot 2017-11-08 at 11 52 05
